components/mediation-connector/org.wso2.carbon.mediation.connector.as4/src/main/java/org/wso2/carbon/mediation/connector/message/util/AS4Utils.java

```diff
@@ -0,0 +1,140 @@
+        }
+        if (messaging.getUserMessage() == null) {
+            throw new AS4Exception("UserMessage element not found", AS4ErrorMapper.ErrorCode.EBMS0001, null);
+        }
+        if (messaging.getUserMessage().getMessageInfo() == null) {
+            throw new AS4Exception("MessageInfo element not found", AS4ErrorMapper.ErrorCode.EBMS0001, null);
+        }
+
+        String messageId = messaging.getUserMessage().getMessageInfo().getMessageId();
+        if (messageId == null || messageId.isEmpty()) {
+            throw new AS4Exception("MessageId element not found", AS4ErrorMapper.ErrorCode.EBMS0001, null);
+        }
+        if (messaging.getUserMessage().getCollaborationInfo() == null) {
+            throw new AS4Exception("CollaborationInfo element not found", AS4ErrorMapper.ErrorCode.EBMS0001, messageId);
+        }
+        String agreementRef = messaging.getUserMessage().getCollaborationInfo().getAgreementRef();
+        if (agreementRef == null || agreementRef.isEmpty()) {
+            throw new AS4Exception("AgreementRef element not found", AS4ErrorMapper.ErrorCode.EBMS0001, messageId);
+        }
+
+        PMode pmode = pmodeRepository.findPModeFromAgreement(agreementRef);
+        if (pmode == null) {
+            throw new AS4Exception("No matching P-Mode found", AS4ErrorMapper.ErrorCode.EBMS0010, messageId);
+        }
+        if(!pmode.getInitiator().getParty().equals(messaging.getUserMessage().getPartyInfo().getFrom().getPartyId().getValue())) {
+            throw new AS4Exception("No matching Initiator Party found in P-Mode", AS4ErrorMapper.ErrorCode.EBMS0010, messageId);
+        }
+        if(!pmode.getResponder().getParty().equals(messaging.getUserMessage().getPartyInfo().getTo().getPartyId().getValue())) {
+            throw new AS4Exception("No matching Responder Party found in P-Mode", AS4ErrorMapper.ErrorCode.EBMS0010, messageId);
+        }
+    }
+
+    /**
+     * Determine if the error need to be notified to the producer end
+     * @param pmode {@link PMode} object
+     * @return true if <ProcessErrorNotifyProducer>true</ProcessErrorNotifyProducer>
+     */
+    public static boolean reportErrorToProducer(PMode pmode) {
+
+        boolean notifyProducer = false;
+        if(pmode.getErrorHandling() != null) {
+            if(pmode.getErrorHandling().getReport() != null) {
+                notifyProducer = pmode.getErrorHandling().getReport().isProcessErrorNotifyProducer();
+            }
+        }
+        return notifyProducer;
+    }
+
+    /**
+     * Determine if the error need to be sent as response to the sending MSH
+     * @param pmode {@link PMode} object
+     * @return true if <AsResponse>true</AsResponse>
+     */
+    public static boolean isSendErrorAsResponse(PMode pmode) {
+
+        boolean notifyProducer = false;
+        if(pmode.getErrorHandling() != null) {
+            if(pmode.getErrorHandling().getReport() != null) {
```

Comment:
Fixed in https://github.com/wso2/carbon-mediation/pull/842/commits/888d7d1b5084b251f5697ea55d3fd77bdf38ebc5